internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// priorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type priorityQueue struct {
+	// queue for wrp messages, ingested by serviceQOS
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *priorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *priorityQueue) Enqueue(msg wrp.Message) {
```

Comment:
if we wanted to enforce a max message payload, then what size limit would we start with and do we want it to be configurable?

Reply:
that's a great point!